File: Transposition/C++/TranspositionTest.cpp
#include "gtest/gtest.h"
#include "Transposition.h"

TEST(TranspositionTest, EncodeDecodeSimple)
{
    const transposition t("HELLO WORLD");
    const std::string encoded = t.encode();
    const std::string decoded = transposition::decode(encoded);

    EXPECT_EQ(encoded, "ODLLLREOHW"); 
    EXPECT_EQ(decoded, "HELLOWORLD");
}

TEST(TranspositionTest, ExactFiveLetters)
{
    const transposition t("ABCDE");
    const std::string encoded = t.encode();
    const std::string decoded = transposition::decode(encoded);

    EXPECT_EQ(encoded, "EDCBA");
    EXPECT_EQ(decoded, "ABCDE");
}

TEST(TranspositionTest, IgnoreNonAlpha)
{
    const transposition t("A!B@C#D$E%");
    const std::string encoded = t.encode();
    const std::string decoded = transposition::decode(encoded);

    EXPECT_EQ(encoded, "EDCBA");
    EXPECT_EQ(decoded, "ABCDE");
}

TEST(TranspositionTest, TwoRowsEncoding)
{
    const transposition t("ABCDEFGHIJ");
    const std::string encoded = t.encode();
    const std::string decoded = transposition::decode(encoded);

    EXPECT_EQ(encoded, "EJDICHBGAF");
    EXPECT_EQ(decoded, "ABCDEFGHIJ");
}

TEST(TranspositionTest, EmptyInput)
{
    const transposition t("");
    const std::string encoded = t.encode();
    EXPECT_EQ(encoded, "");

    const std::string decoded = transposition::decode(encoded);
    EXPECT_EQ(decoded, "");
}

TEST(TranspositionTest, LessThanFiveLetters)
{
    const transposition t("AB");
    const std::string encoded = t.encode();
    EXPECT_EQ(encoded, "");

    const std::string decoded = transposition::decode(encoded);
    EXPECT_EQ(decoded, "");
}

TEST(TranspositionTest, MixedCaseInput)
{
    const transposition t("AbCdEfGhIj");
    const std::string encoded = t.encode();
    const std::string decoded = transposition::decode(encoded);

    EXPECT_EQ(encoded, "EJDICHBGAF");
    EXPECT_EQ(decoded, "ABCDEFGHIJ");
}

File: Transposition/Rust/transposition_test.rs
#[cfg(test)]
mod tests {
    use super::transposition::Transposition;

    #[test]
    fn test_new_creates_instance_with_input() {
        let t = Transposition::new("hello".to_string());
        assert_eq!(t.input, "hello");
    }

    #[test]
    fn test_encode_basic_input() {
        let t = Transposition::new("hello world".to_string());
        let encoded = t.encode();
        assert!(!encoded.is_empty());
      
    }

    #[test]
    fn test_encode_removes_non_alphabetic() {
        let t = Transposition::new("a!b@c#d$e%f^".to_string());
        let encoded = t.encode();
        assert!(!encoded.contains('!'));
        assert!(!encoded.contains('@'));
        assert!(encoded.chars().all(|c| c.is_alphabetic()));
    }

    #[test]
    fn test_encode_uppercase_output() {
        let t = Transposition::new("abcde".to_string());
        let encoded = t.encode();
        assert!(encoded.chars().all(|c| c.is_uppercase()));
    }

    #[test]
    fn test_encode_empty_string() {
        let t = Transposition::new("".to_string());
        let encoded = t.encode();
        assert_eq!(encoded, "");
    }

    #[test]
    fn test_decode_reverses_encoding() {
        let input = "thequickbrownfox";
        let t = Transposition::new(input.to_string());
        let encoded = t.encode();
        let decoded = t.decode(&encoded);
        for c in input.chars().filter(|c| c.is_alphabetic()) {
            assert!(decoded.contains(c.to_ascii_uppercase()));
        }
    }

    #[test]
    fn test_encode_non_multiple_of_5_characters() {
        let t = Transposition::new("abcdefg".to_string());
        let encoded = t.encode();
        assert!(encoded.len() >= 7);
    }
}
